nni/retiarii/experiment/config/experiment_config.py

```diff
@@ -46,15 +51,25 @@ def _canonicalize(self, _parents):
         # TODO: maybe we should also allow users to specify trial_code_directory
         if str(self.trial_code_directory) != '.' and not os.path.isabs(self.trial_code_directory):
             raise ValueError(msg.format('trial_code_directory', self.trial_code_directory))
-        if self.trial_command != '_reserved' and \
-            not self.trial_command.startswith('python3 -m nni.retiarii.trial_entry '):
+
+        trial_command_tmpl = '{envs} {python} -m nni.retiarii.trial_entry {execution_engine}'
+        if self.trial_command != '_reserved' and '-m nni.retiarii.trial_entry' not in self.trial_command:
             raise ValueError(msg.format('trial_command', self.trial_command))
 
         if isinstance(self.execution_engine, str):
             self.execution_engine = execution_engine_config_factory(self.execution_engine)
-        if self.execution_engine.name in ('py', 'base', 'cgo'):
-            # TODO: replace python3 with more elegant approach
-            # maybe use sys.executable rendered in trial side (e.g., trial_runner)
-            self.trial_command = 'python3 -m nni.retiarii.trial_entry ' + self.execution_engine.name
+
+        _trial_command_params = {
+            # Default variables
+            'envs': '',
+            # TODO: maybe use sys.executable rendered in trial side (e.g., trial_runner)
+            'python': sys.executable,
+            'execution_engine': self.execution_engine.name,
+
+            # Overridden by new
+            **(self._trial_command_params or {})
```

Comment:
according to line 73, this line does not take any effect. so why add it?

Reply:
Why no effect?